### ISSUE TITLE
Garbage collector `min_age` should start when SSTs become inactive

### ISSUE THREAD

**Issue by criccomini** [slatedb/slatedb#204]
The garbage collector's `min_age` setting for SSTs currently just looks at when an object's last-modify-time is more than N seconds old. I think for WAL SSTs this is fine. But for compacted SSTs, which can be long-lived, this results an in odd behavior:

An SST that's been active (in the manifest) for more than 24h will get GC'd almost as soon as it's removed from the manifest. Consider the case where an SST in an old/big sorted run has been around for a three days. The compactor finally reaches the point where it compacts out that sorted run and removes it from the manifest. The GC's next poll interval (defaults to 5 minutes) will see the SSTs in this sorted run are > 24h old and delete it. This is pretty abrupt, though it shouldn't cause any problems.

The question is: do we want `min_age` clock to start based on when the SSTs are written or when they become inactive.

If we want the `min_age` clock to start when SSTs become inactive, we could determine their inactivity based on the manifest history we have. What I'm thinking is the GC could do something like this:

1. List all manifests (sorted from old to new).
2. Iterate through each manifest, detecting when L0/compacted SSTs become inactive (when they're no longer referenced).
3. Set the expiry timestamp for each detected SST based on the first manifest timestamp where they're no longer referenced.
4. Delete all SSTs with an expiry timestamp > 24h or that are unreferenced by any manifest and have a last modify time > 24h.
6. Delete all manifests that are not part of any snapshot and reference _only_ SSTs that have all been deleted.

We could theoretically apply similar logic to the `wal` SSTs starting from `last_compacted_wal_id`'s last modify time rather than each WAL SST's last modify time. If we make the above changes, it's probably worth doing this to make things consistent.

This change makes the manifest deletions a little less predictable since we'd have to keep manifests around until all their referenced SSTs have been deleted. I don't see an issue with that, though. Practically speaking, the manifest and compacted SSTs default to the same GC settings, so all of a manifest's SST should be deleted at about the same time the manifest expires anyway.

**Comment by rodesai** [comment]
> for each detected SST

I think there's two ways we can detect SSTs with this approach - by actually listing them all like we do today, or by looking at the SSTs reported in the manifests. The latter is cheaper, but won't include SSTs that were left behind by writers/compactors that never got a chance to write them to a manifest. So, I think we should still be listing all the object store SSTs when looking for candidates to delete.

This came up in a conversation w/ @agavra.

**Comment by criccomini** [comment]
Correct. I didn't mention it explicitly, but you would need to do a full list to complete this part:

> or that are unreferenced by any manifest and have a last modify time > 24h.

**Comment by rodesai** [comment]
Documenting one option for supporting this that builds on the proposal in https://github.com/slatedb/slatedb/pull/278. The compactor could leave a checkpoint for the manifest _before_ applying it's updates. The checkpoint could have an expiry time that matches the time that we want to retain SSTs for after they have been removed from the db. The GC will delete this checkpoint once it's expired and then no longer include the referenced manifest in the set of valid manifests that it includes when building up the active SST list. This is basically the same as what's proposed in the original issue description, except it restricts the set of manifest versions to search to those immediately preceding a change to the db that removes SSTs.

**Comment by criccomini** [comment]
Ohhhh, this is a clever approach. It's much simpler to implement! I like it.

**Comment by vigneshc** [comment]
+1 on the approach shared above. It requires #319 and #311  to be implemented first.
